### PR TITLE
 Improved presentation flexibility

### PR DESCRIPTION
**Pull request by acrusage-iaik**
This pull requests adds logic to handle a presentations more diligently, but this unfortunately leads to some changes in semantics expected from the "createPresentation" function before that. I hope the changes are reasonable, I tried following the guide from PresentationExchange at:
https://identity.foundation/presentation-exchange/spec/v2.0.0/#input-evaluation

I tried to stay conform with openId4vp, but I hope that we don't need to support other presentation specifications?

Semantic changes to be approved:
### Presentation Submissions now only contain matches from input descriptors.

Previously a verifiable presentation contained all the verifiable credentials with the required credential scheme, but not anymore.
Each input descriptor matches to exactly one verifiable credential, and that verifiable credential must satisfy all constraint fields.
This may actually break some things, as it was previously possible to present multiple AtomicCredentials at once, but after this change each AtomicCredential attribute needs to be requested separately (in separate input descriptors).
-> result.vp.verifiableCredentials now always has size 1!
Impacted Tests (among others):
- AgentTest."valid presentation is valid": change last "shouldHaveSize" to "1"
- ValidatorVpTest."valid parsed presentation should separate revoked and valid credentials": Change last "shouldHaveSize" to "issuedCredential.successful.size"

### Presentation Submissions where no fields are requested don't return any fields

Previously it seems like not requesting any claims should return all available claims, but not anymore.
Impacted Tests (among others):

- OidcSiopIsoProtocolTest."test with Fragment for mDL": request an attribute to satisfy the "document.validItems.shouldNotBeEmpty()" assertion
- OidcSiopIsoProtocolTest."test with Fragment for custom attributes": request an attribute to satisfy the "document.validItems.shouldNotBeEmpty()" assertion

### Scope value semantics have somewhat changed

Previously the scope values were used to deduce the requested credential schemes, but not anymore.
As per openid4vp the scope values have the following semantics:

    Such a scope value MUST be an alias for a well-defined Presentation Definition that will be referred to in the presentation_submission response parameter.
    https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#section-5.3-2

scope values are therefore only taken into consideration now in case neither presentationDefinition nor presentationDefinitionUri return a presentation definition.

### Users can now define a path authorization validator 
This can mainly be done by providing a default authorizer in the constructor of HolderAgent.

### CODE REVIEW

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt
Comment (nodh) [inline vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt:10]: why not use the artifact of https://github.com/a-sit-plus/eu-pid-credential?

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopIsoProtocolTest.kt
Comment (nodh) [inline vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopIsoProtocolTest.kt:103]: why not use a constant here?

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (nodh) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:138]: imho that's not part of the logic in vclib, especially not at the holder interface

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (nodh) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:146]: is this TODO still relevant?

**Review by nodh — CHANGES_REQUESTED** [review]
Just a few remarks, since this is a draft still.
But wouldn't it be cleaner to create a new, separate, library for JSON Path processing? There seems to be a lot of classes involved.

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt
Comment (acrusage) [reply vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt:10]: Should this really be a dependency of the vclib?
What if the EUDIW PID credential changes in the future? This may break the tests, even though the tested functionality still remains correct?
Maybe it would be better to just rename it to TestCredential or something like that?

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopIsoProtocolTest.kt
Comment (acrusage) [reply vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopIsoProtocolTest.kt:103]: AtomicAttribute2023 does not specify any attribute constants

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (acrusage) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:138]: In my opinion it would be actor performing the selective disclosure procedure that must be the one who is aware of the verifiers permissions in accessing them..
Alternatively there would need to be a separate check before or after the selective disclosure has been performed, and in this case the TODO would have to be moved somewhere else?

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (acrusage) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:146]: I am not sure to be honest...
I think it has already been there before, and I am not aware of its relevance as there is not more information on why the TODO exists in the first place

**Comment by acrusage** [comment]
> Just a few remarks, since this is a draft still. But wouldn't it be cleaner to create a new, separate, library for JSON Path processing? There seems to be a lot of classes involved.

Thanks!
Yes, that would be a better approach, but I'm unfortunately not yet familiar with creating KMP libraries or publishing..

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt
Comment (nodh) [reply vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt:10]: Then why not rename it to "TestCredential", or if it tests something EU-PID-specific, move the test to the eu-pid repository? But imho depending on version 1.0.0 of that EU PID artifact isn't that bad, because future updates to that credential should remain backwards-compatible, or be released as 2.0.0, so it won't be picked up by vclib.

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (nodh) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:138]: You're right, there should be something like a verification method (or function pointer) in the holder class, that Wallets need to implement (with the default implementation to allow all requests).

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (nodh) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:146]: You're right, that's on my side ...

**Comment by acrusage-iaik** [comment]
@nodh @JesusMcCloud PR is now ready for review & merge

File: conventions-vclib/src/main/kotlin/VcLibConventions.kt
Comment (acrusage-iaik) [inline conventions-vclib/src/main/kotlin/VcLibConventions.kt:1]: Is it normal that these changes are part of a pull request on this repository? Shouldn't this result in changes to the conventions submodule?

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt
Comment (acrusage-iaik) [reply vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/EudiwPidCredentialScheme.kt]: Has been fixed somewhere along the way, not sure exactly where

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (acrusage-iaik) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:138]: Todo for the future

File: conventions-vclib/src/main/kotlin/VcLibConventions.kt
Comment (JesusMcCloud) [reply conventions-vclib/src/main/kotlin/VcLibConventions.kt:1]: I need to evaluate whether we still need the VcLibConventions, now that our generic conventions plugin supports version catalogs. I'm pretty sure that @nodh's suspicion that we should can it is correct. for this PR, though we should leave it like this.
I've assigned myself issue #59 for that.

**Review by JesusMcCloud — COMMENTED** [review]
externalisation of jsonpath looks good, but is should we not wait for the next release of the library before we continue here? (I'll try to do that today)

**Comment by acrusage-iaik** [comment]
> externalisation of jsonpath looks good, but is should we not wait for the next release of the library before we continue here? (I'll try to do that today)

This should not be an issue for us
The next release only makes changes to function extensions, which we will most likely not use for now

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (nodh) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt:138]: see https://github.com/a-sit-plus/kmm-vc-library/issues/60

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopInteropTest.kt
Comment (nodh) [inline vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopInteropTest.kt:143]: Why did this change?

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt
Comment (nodh) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt:235]: I would not put a `data class` inside a method. Maybe move out some parts of this large method?

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt
Comment (nodh) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt:477]: maybe put that into a separate file, so that this file here doesn't get over 500 lines long (which is way too much)

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/data/dif/InputEvaluator.kt
Comment (nodh) [inline vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/data/dif/InputEvaluator.kt:16]: Please use `/**` and also move the comment in `//` into here

File: vclib/src/commonMain/resources/grammar/JsonPath.abnf
Comment (nodh) [inline vclib/src/commonMain/resources/grammar/JsonPath.abnf:1]: isn't this file part of the JSONPath library now? Same for the `g4` ones

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/agent/AgentSdJwtTest.kt
Comment (nodh) [inline vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/agent/AgentSdJwtTest.kt:51]: Code here seems duplicated with functions below

**Review by nodh — CHANGES_REQUESTED** [review]
Great work, just some minor "editorial" changes requested

File: vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopInteropTest.kt
Comment (acrusage-iaik) [reply vclib-openid/src/commonTest/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopInteropTest.kt:143]: I just did a new presentation and copied the new parameters there

It seems like their presentation request changed from back then to now, so I figured it would be a good idea to update this

**Review by JesusMcCloud — APPROVED** [review]
approved, pending the change requests by @nodh

File: vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/agent/AgentSdJwtTest.kt
Comment (acrusage-iaik) [reply vclib/src/commonTest/kotlin/at/asitplus/wallet/lib/agent/AgentSdJwtTest.kt:66]: fixed in commit 950b306a68c5b7a9af3591d61dead5fe364bd94c

File: vclib/src/commonMain/resources/grammar/JsonPath.abnf
Comment (acrusage-iaik) [reply vclib/src/commonMain/resources/grammar/JsonPath.abnf]: fixed in 3e4c8cd54334e7b7ed0bb56817ae64d8ed0d8d0d

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/data/dif/InputEvaluator.kt
Comment (acrusage-iaik) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/data/dif/InputEvaluator.kt]: fixed in 3e4c8cd54334e7b7ed0bb56817ae64d8ed0d8d0d

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt
Comment (acrusage-iaik) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt]: fixed in 3e4c8cd54334e7b7ed0bb56817ae64d8ed0d8d0d

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt
Comment (acrusage-iaik) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/HolderAgent.kt]: Moved to `vclib/.../data/CredentialToJsonConverter` in commit 3e4c8cd54334e7b7ed0bb56817ae64d8ed0d8d0d

After refactoring for comment https://github.com/a-sit-plus/kmm-vc-library/pull/53#discussion_r1593812217 it got way too big though, so I also moved the `create*Presentation-files` to their own class: `vclib/.../agent/VerifiablePresentationFactory`
I'm not so sure whether this makes the code too loose and hard to grasp though, and I'm also unsure about the location and the name of the new class

**Comment by acrusage-iaik** [comment]
> Great work, just some minor "editorial" changes requested

Unfortunately I also implemented other things while doing the refactoring to make the method smaller, so I guess I'll have to re-request a review..
Sorry about that!

**Comment by JesusMcCloud** [comment]
To be frank: I lack the context to judge these last changes.

**Review by nodh — CHANGES_REQUESTED** [review]
Great work! Some minor questions remain

File: vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt
Comment (acrusage-iaik) [reply vclib/src/commonMain/kotlin/at/asitplus/wallet/lib/agent/Holder.kt]: fixed/reworked in commit 79ba0e4f03b62d8f3e5fd2fca38190a3b113e2a9

File: vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt
Comment (acrusage-iaik) [reply vclib-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidc/OidcSiopWallet.kt:69]: added tests in commit e73130c89d6226fd87bb80c5a83d755d9792b851

**Comment by acrusage-iaik** [comment]
> To be frank: I lack the context to judge these last changes.

I removed work towards submission requirements as it wasn't polished at all, maybe it's easier to review now?

**Review by nodh — APPROVED** [review]
Let's merge everything together for vclib 3.7.0 and see how it works out :-)

**Comment by JesusMcCloud** [comment]
Well now that its out, we could just bump jsonpath to 2.0.0?

**Pull request merged**
